docs/internal/types.md

```diff
@@ -0,0 +1,236 @@
+
+Capabilities offer language and runtime guarantees:
+ * `iso`, or `isolated`, means this is a pointer to an exclusive area of memory that only this pointer has access to in the entire program.
+ * `mut`, or `mutable`, means this object can be changed at will, which means it must be inside an area of memory you have exclusive access to, ie. traced back to an `iso` you hold the rights.
+ * `imm`, or `immutable`, means no one can write to, so anyone can read from at any time. You can make `iso` regions immutable but not the other way round, as there's no guarantees who's still reading from it.
+ * `readonly` is a local capability that is used when you want to make sure an external function can't modify, but you still hold a mutable reference.
+
+Note that type capabilities, unions and intersections are types themselves.
+So `(A & imm)` is completely different from `(A & mut)` or `(A & iso)`.
+Specifically, you cannot convert `mut` to `imm` (as you would with a `const` cast in C++), or vice versa.
+You can freeze an `iso` region into `imm`, but unfreeze an `imm` region.
+Most importantly, capabilities are not type traits like C++'s `const int * const`, they have specific non-trivial runtime semantics.
+
+The type system guarantees that:
+ * Structural soundness: Interfaces and classes form valid concrete types, operations and functions take in the correct (sometimes inferred) types.
+ * Capability guarantees are followed: The compiler checks operations on types, the runtime guarantees the semantics is followed.
+ * Runtime safety: A program that compiles is a program that runs safely, without crashes, traps or unhandled exceptions, including asynchronous behaviour, but not necessarily producing the right result or doing what you expect.
+
+## Classes and Interfaces
+
+Classes and interfaces have a similar structure:
+ * Fields: elements of a specific type that have a runtime space allocated.
+ * Methods: functions that belong to a type.
+
+The declaration is trivial:
+```ts
+// Helps receivers know they can convert the object to a string representation
+interface Serialisable
+{
+  // All derived classes will have the common field
+  var common : F64;
+
+  // Derived classes must implement this method
+  toString() : String;
+};
+
+class MyType : Serialisable
+{
+  // Inherits `common`
+  let field : I32;
+
+  // Some other method
+  innerSum(x : I32) : I32
+  {
+    let res : I32 = x + field;
+    res
+  }
+
+  // Implements `toString`
+  toString() : String
+  {
+    ...
+  }
+};
+```
+
+### Fields
+
+Define objects that belong to the class and will be instantiated (ie. use memory) on every instance of the class.
+
+Fields can be initialised in two ways:
+ * `var`: Must be initialised on construction, but can change after.
+ * `let`: Must be initialised on construction, and cannot change later.
+
+They can also be a private member, if its name starts with `_`.
+
+They can also have a keyword `embed`, which means ...
```

Comment:
Missing content here. The current writing makes it sound like I can do `let embed foo`.

Reply:
On purpose, I'm still confused as to what `embed` will actually mean in the language.

Reply:
Syntactically, I believe `embed` is used in place of `let` or `var`, though I'm not sure what this means with respect to capabilities (`embed thing : T & imm` probably doesn't make sense because you can't share a pointer to `thing` between `cown`s, can you `embed thing : T & readonly`?), but assuming that's sorted we will be able to do the equivalent of `var` vs `let` with capabilities on the `embed` field.

Semantically, it just means put the fields of the object here.

Reply:
Right, but I still have to explain where do we put the field when it's not an `embed`. ie. where is `here`? Is this like `static` fields in C++?

Reply:
All fields are stored in the object (unless they're class fields rather than instance fields [static fields, in C++ / Java terminology] - I don't know if we have syntax for that yet?

The difference between `let` / `var` and `embed` is the difference between a pointer and a value.  Here's some C++:

```c++
struct Point
{
  double x, y;
};

struct Example
{
  // let l : Point & mut
  const Point *l;
  // let li : Point & imm
  const Point * const li; // I think - C / C++ `const` placement confuses me
  // var v : Point
  Point *v;
  // embed e : Point
  Point e;
};
```

Reply:
Ah! So much easier to understand when you put it into code. :)

So, this means that the structure layout of `Example` would be something like:
```
{ USize, USize, USize, F64, F64 }
```

And the compiler is in charge of encoding access (direct or indirect), but also making sure the field indexes are packed, aligned and available as an interface type that they implement (ie. same offsets?), according to the semantics of the language (and their modifiers).

I'm not sure your `const` comparison has the same semantics, so I won't include that in the document to avoid misleading readers. But I've added the description and an example in Verona.